xla/hlo/analysis/hlo_replication_analysis.cc

```diff
@@ -120,6 +124,8 @@ HloReplicationAnalysis::DetermineHloInstructionIsReplicated(
         if ((cross_partition_spmd && replicated_across_partitions) ||
             (!cross_partition_spmd && replicated_across_replicas)) {
           return HloReplication::ReplicatedOnAllDevices();
+        } else if (support_partial_replication) {
+          return HloReplication::PartiallyReplicated(device_sets);
```

Comment:
The device IDs in device_sets should be partition IDs if cross_partition_spmd=true or replica IDs if cross_partition_spmd=false. But you're adding flattened IDs here. 

Convert to replica or partition IDs, by dividing or modding by num_partitions. This isn't strictly correct: You should also ensure that if cross_partition_spmd=true, all replicas have the particular partial replication, and simiarly if cross_partition_spmd=false, all partitions have the particular partial replication. But we already have a similar bug in the original code where we might incorrectly return ReplicatedOnAllDevices since we don't check individual replicas/partitions. Feel free to fix this here and the original code, or not. If not, I will fix this up myself later.

It might be useful to support flattened IDs here, but you'll have to provide an API to this class to allow the user to query either flattened IDs or partitions/replica IDs. I think we initially should not add an API, and add it later if necessary.